{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In this file we try to identify the best parameters for the amplicone design tool given known parameters\n",
    "#%%\n",
    "from functools import partial\n",
    "from random import choices, randint, randrange, random, sample\n",
    "from typing import List, Optional, Callable, Tuple\n",
    "import numpy as np\n",
    "# from geneticalgorithm import geneticalgorithm as ga\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from tqdm import tqdm\n",
    "import time\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "from Bio import SeqIO\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "import primer3\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "from primer3 import calc_heterodimer\n",
    "from primer3 import bindings"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tm: 30.40290207197791\n",
      "GC content: 50.0\n"
     ]
    }
   ],
   "source": [
    "reference_genome= '/mnt/storage10/lwang/Projects/Amplicone_design_tool/MTB-h37rv_asm19595v2-eg18.fa'\n",
    "\n",
    "def calculate_gc_content(sequence):\n",
    "    \"\"\"\n",
    "    Calculate the percentage of G and C nucleotides in a DNA sequence.\n",
    "\n",
    "    Args:\n",
    "        sequence (str): DNA sequence string.\n",
    "\n",
    "    Returns:\n",
    "        float: Percentage of G and C nucleotides in the sequence.\n",
    "    \"\"\"\n",
    "    gc_count = 0\n",
    "    total_count = 0\n",
    "\n",
    "    for nucleotide in sequence:\n",
    "        if nucleotide.upper() in ['G', 'C']:\n",
    "            gc_count += 1\n",
    "        if nucleotide.upper() in ['A', 'T', 'G', 'C']:\n",
    "            total_count += 1\n",
    "\n",
    "    gc_percentage = (gc_count / total_count) * 100\n",
    "    return gc_percentage\n",
    "\n",
    "def extract_sequence_from_fasta(start_pos, end_pos, padding = 150, fasta_file= '/mnt/storage10/lwang/Projects/Amplicone_design_tool/MTB-h37rv_asm19595v2-eg18.fa', sequence_id='Chromosome'):\n",
    "    \"\"\"\n",
    "    Extracts a subsequence from a FASTA file based on the given sequence ID, start position, and end position.\n",
    "    \"\"\"\n",
    "    # Iterate over the sequences in the FASTA file\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        # Check if the current sequence ID matches the desired sequence ID\n",
    "        \n",
    "        if record.id == sequence_id:\n",
    "            # Extract the subsequence based on the start and end positions\n",
    "            subsequence = record.seq[start_pos-padding:end_pos+padding]\n",
    "            return str(subsequence)  # Return the subsequence as a string\n",
    "    # If the sequence ID is not found, return None\n",
    "    return None\n",
    "\n",
    "def complement_sequence(seq):\n",
    "    complement = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n",
    "    return \"\".join(complement[base] for base in seq)\n",
    "\n",
    "def reverse_complement_sequence(seq):\n",
    "    complement = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n",
    "    reverse_seq = seq[::-1]\n",
    "    return \"\".join(complement[base] for base in reverse_seq)\n",
    "\n",
    "def find_sequence_location(query_sequence, position=0, fasta_file=reference_genome):\n",
    "    closest_start = float('inf')\n",
    "    closest_end = float('inf')\n",
    "    closest_record = None\n",
    "\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        sequence = str(record.seq)\n",
    "        index = sequence.find(query_sequence)\n",
    "        \n",
    "        while index != -1:\n",
    "            start_index = index\n",
    "            end_index = index + len(query_sequence)\n",
    "            \n",
    "            if abs(position - start_index) < abs(position - closest_start):\n",
    "                closest_start = start_index\n",
    "                closest_end = end_index\n",
    "                closest_record = record.id\n",
    "\n",
    "            index = sequence.find(query_sequence, index + 1)\n",
    "\n",
    "    if closest_record is not None:\n",
    "        return closest_start, closest_end\n",
    "    else:\n",
    "        print(\"Primer Sequence not found in FASTA file.\")\n",
    "        return 0,0\n",
    "    \n",
    "from difflib import SequenceMatcher\n",
    "def check_similarity(seq1, seq2, threshold=0.3):\n",
    "    #Check if two sequences look similar based on their similarity score.\n",
    "    # Parameters:\n",
    "    # - seq1: First sequence (string)\n",
    "    # - seq2: Second sequence (string)\n",
    "    # - threshold: Similarity threshold (float between 0 and 1)\n",
    "    # Returns:\n",
    "    # - True if the sequences look similar, False otherwise\n",
    "    \n",
    "    similarity_score = SequenceMatcher(None, seq1, seq2).ratio()\n",
    "    if similarity_score >= threshold:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "def calculate_tm_gc(primer_sequence):\n",
    "    # Calculate Tm\n",
    "    tm = MeltingTemp.Tm_GC(primer_sequence)\n",
    "    # Calculate GC content\n",
    "    gc_count = primer_sequence.count('G') + primer_sequence.count('C')\n",
    "    gc_content = (gc_count / len(primer_sequence)) * 100\n",
    "\n",
    "    return tm, gc_content\n",
    "# Example usage\n",
    "primer = 'ATCGATCGATCG'\n",
    "tm, gc_content = calculate_tm_gc(primer)\n",
    "print('Tm:', tm)\n",
    "print('GC content:', gc_content)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# loading in the data of known primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "known_primers = pd.read_excel('/mnt/storage10/lwang/Projects/Amplicone_design_tool/amplicon_TB_2023_V2.xlsx', sheet_name='primers', header=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "known_primers = known_primers[['Drug','primer name', 'forward', 'reverse', 'Amplicon size', 'Start', 'End']]\n",
    "known_primers = known_primers.iloc[:39,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug</th>\n",
       "      <th>primer name</th>\n",
       "      <th>forward</th>\n",
       "      <th>reverse</th>\n",
       "      <th>Amplicon size</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Fluoroquinolones</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>AGAGTTGGTGCGGCGTAA</td>\n",
       "      <td>GCCACTTGAGTTTGTACAGC</td>\n",
       "      <td>449.0</td>\n",
       "      <td>6511.0</td>\n",
       "      <td>6959.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fluoroquinolones</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>GTTGACATCGAGCAGGAGAT</td>\n",
       "      <td>AAATCGACTGTCTCCTCGTC</td>\n",
       "      <td>404.0</td>\n",
       "      <td>7353.0</td>\n",
       "      <td>7756.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rifampicin</td>\n",
       "      <td>rpoB_1</td>\n",
       "      <td>GGTGGCCTGGAAGAGGT</td>\n",
       "      <td>AGAACCCGAACCGCTCG</td>\n",
       "      <td>514.0</td>\n",
       "      <td>760038.0</td>\n",
       "      <td>760551.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rifampicin</td>\n",
       "      <td>rpoB_2</td>\n",
       "      <td>CCCCGACCAAAGAGTCA</td>\n",
       "      <td>CTCACGTGACAGACCGC</td>\n",
       "      <td>542.0</td>\n",
       "      <td>760645.0</td>\n",
       "      <td>761186.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rifampicin</td>\n",
       "      <td>rifampicin</td>\n",
       "      <td>TGCTCACCTCGATCCACATC</td>\n",
       "      <td>ATCTTGCGTTTCTGAGCCAC</td>\n",
       "      <td>400.0</td>\n",
       "      <td>762025.0</td>\n",
       "      <td>762424.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Drug primer name               forward               reverse  \\\n",
       "0  Fluoroquinolones        gyrB    AGAGTTGGTGCGGCGTAA  GCCACTTGAGTTTGTACAGC   \n",
       "1  Fluoroquinolones        gyrA  GTTGACATCGAGCAGGAGAT  AAATCGACTGTCTCCTCGTC   \n",
       "2        Rifampicin      rpoB_1     GGTGGCCTGGAAGAGGT     AGAACCCGAACCGCTCG   \n",
       "3        Rifampicin      rpoB_2     CCCCGACCAAAGAGTCA     CTCACGTGACAGACCGC   \n",
       "4        Rifampicin  rifampicin  TGCTCACCTCGATCCACATC  ATCTTGCGTTTCTGAGCCAC   \n",
       "\n",
       "   Amplicon size     Start       End  \n",
       "0          449.0    6511.0    6959.0  \n",
       "1          404.0    7353.0    7756.0  \n",
       "2          514.0  760038.0  760551.0  \n",
       "3          542.0  760645.0  761186.0  \n",
       "4          400.0  762025.0  762424.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "known_primers.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# getting the tm, gc and primer size range of hte known primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PRIMER_INTERNAL_OPT_SIZE: 6\n",
      "PRIMER_INTERNAL_MAX_POLY_X: 0\n",
      "PRIMER_MAX_POLY_X: 0\n",
      "PRIMER_GC_CLAMP: 0.6666666666666666\n",
      "PRIMER_MAX_END_GC: 1.0\n"
     ]
    }
   ],
   "source": [
    "def get_primer_parameters(primers, amplified_sequence):\n",
    "    \"\"\"Returns the PRIMER_INTERNAL_OPT_SIZE, PRIMER_INTERNAL_MAX_POLY_X,\n",
    "    PRIMER_MAX_POLY_X, PRIMER_GC_CLAMP, and PRIMER_MAX_END_GC parameters for the\n",
    "    given primers and amplified sequence.\"\"\"\n",
    "\n",
    "    primer_internal_opt_size = min(len(primer) for primer in primers)\n",
    "    primer_internal_max_poly_x = max(\n",
    "        sum(1 for base in primer if base == base * 2) for primer in primers)\n",
    "    primer_max_poly_x = max(\n",
    "        sum(1 for base in primer if base == base * 3) for primer in primers)\n",
    "    primer_gc_clamp = min(\n",
    "        sum(1 for base in primer if base == \"G\" or base == \"C\") / len(primer)\n",
    "        for primer in primers)\n",
    "    primer_max_end_gc = max(\n",
    "        sum(1 for base in primer[-3:] if base == \"G\" or base == \"C\") / 3\n",
    "        for primer in primers)\n",
    "\n",
    "    return primer_internal_opt_size, primer_internal_max_poly_x, primer_max_poly_x, primer_gc_clamp, primer_max_end_gc\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    primers = [\"ATGCGG\", \"CTAGGG\"]\n",
    "    amplified_sequence = \"ATGCGGCTAGGG\"\n",
    "\n",
    "    primer_parameters = get_primer_parameters(primers, amplified_sequence)\n",
    "\n",
    "    print(\"PRIMER_INTERNAL_OPT_SIZE:\", primer_parameters[0])\n",
    "    print(\"PRIMER_INTERNAL_MAX_POLY_X:\", primer_parameters[1])\n",
    "    print(\"PRIMER_MAX_POLY_X:\", primer_parameters[2])\n",
    "    print(\"PRIMER_GC_CLAMP:\", primer_parameters[3])\n",
    "    print(\"PRIMER_MAX_END_GC:\", primer_parameters[4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "highest_tm: 58.81956873864458\n",
      "lowest_tm: 47.06956873864456\n",
      "highest_gc: 68.42105263157895\n",
      "lowest_gc: 45.0\n",
      "longest_primer: 24\n",
      "shortest_primer: 17\n"
     ]
    }
   ],
   "source": [
    "highest_tm = 0\n",
    "lowest_tm = 100\n",
    "highest_gc = 0\n",
    "lowest_gc = 100\n",
    "longest_primer = 0\n",
    "shortest_primer = 100\n",
    "\n",
    "PRIMER_INTERNAL_OPT_SIZE = []\n",
    "PRIMER_INTERNAL_MAX_POLY_X = []\n",
    "PRIMER_MAX_POLY_X = []\n",
    "PRIMER_GC_CLAMP = []\n",
    "PRIMER_MAX_END_GC = []\n",
    "\n",
    "\n",
    "for i, row in known_primers.iterrows():\n",
    "    try:\n",
    "        tm, gc_content = calculate_tm_gc(row['forward'])\n",
    "    except:\n",
    "        print(row['forward'])\n",
    "    if tm > highest_tm:\n",
    "        highest_tm = tm\n",
    "    if tm < lowest_tm:\n",
    "        lowest_tm = tm\n",
    "    if gc_content > highest_gc:\n",
    "        highest_gc = gc_content\n",
    "    if gc_content < lowest_gc:\n",
    "        lowest_gc = gc_content\n",
    "        \n",
    "    tm, gc_content = calculate_tm_gc(row['reverse'])\n",
    "    if tm > highest_tm:\n",
    "        highest_tm = tm\n",
    "    if tm < lowest_tm:\n",
    "        lowest_tm = tm\n",
    "    if gc_content > highest_gc:\n",
    "        highest_gc = gc_content\n",
    "    if gc_content < lowest_gc:\n",
    "        lowest_gc = gc_content\n",
    "        \n",
    "    if len(row['forward']) > longest_primer:\n",
    "        longest_primer = len(row['forward'])\n",
    "    if len(row['forward']) < shortest_primer:\n",
    "        shortest_primer = len(row['forward'])\n",
    "    if len(row['reverse']) > longest_primer:\n",
    "        longest_primer = len(row['reverse'])\n",
    "    if len(row['reverse']) < shortest_primer:\n",
    "        shortest_primer = len(row['reverse'])\n",
    "\n",
    "    primer_parameters = get_primer_parameters([row['forward'],row['reverse']], extract_sequence_from_fasta(int(row['Start']-5), int(row['End']), padding = 0, fasta_file= '/mnt/storage10/lwang/Projects/Amplicone_design_tool/MTB-h37rv_asm19595v2-eg18.fa', sequence_id='Chromosome')):\n",
    "\n",
    "    print(\"PRIMER_INTERNAL_OPT_SIZE:\", primer_parameters[0])\n",
    "    print(\"PRIMER_INTERNAL_MAX_POLY_X:\", primer_parameters[1])\n",
    "    print(\"PRIMER_MAX_POLY_X:\", primer_parameters[2])\n",
    "    print(\"PRIMER_GC_CLAMP:\", primer_parameters[3])\n",
    "    print(\"PRIMER_MAX_END_GC:\", primer_parameters[4])\n",
    "\n",
    "print('highest_tm:', highest_tm)\n",
    "print('lowest_tm:', lowest_tm)\n",
    "print('highest_gc:', highest_gc)\n",
    "print('lowest_gc:', lowest_gc)\n",
    "print('longest_primer:', longest_primer)\n",
    "print('shortest_primer:', shortest_primer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PRIMER_INTERNAL_OPT_SIZE: 6\n",
      "PRIMER_INTERNAL_MAX_POLY_X: 0\n",
      "PRIMER_MAX_POLY_X: 0\n",
      "PRIMER_GC_CLAMP: 0.6666666666666666\n",
      "PRIMER_MAX_END_GC: 1.0\n"
     ]
    }
   ],
   "source": [
    "def get_primer_parameters(primers, amplified_sequence):\n",
    "    \"\"\"Returns the PRIMER_INTERNAL_OPT_SIZE, PRIMER_INTERNAL_MAX_POLY_X,\n",
    "    PRIMER_MAX_POLY_X, PRIMER_GC_CLAMP, and PRIMER_MAX_END_GC parameters for the\n",
    "    given primers and amplified sequence.\"\"\"\n",
    "\n",
    "    primer_internal_opt_size = min(len(primer) for primer in primers)\n",
    "    primer_internal_max_poly_x = max(\n",
    "        sum(1 for base in primer if base == base * 2) for primer in primers)\n",
    "    primer_max_poly_x = max(\n",
    "        sum(1 for base in primer if base == base * 3) for primer in primers)\n",
    "    primer_gc_clamp = min(\n",
    "        sum(1 for base in primer if base == \"G\" or base == \"C\") / len(primer)\n",
    "        for primer in primers)\n",
    "    primer_max_end_gc = max(\n",
    "        sum(1 for base in primer[-3:] if base == \"G\" or base == \"C\") / 3\n",
    "        for primer in primers)\n",
    "\n",
    "    return primer_internal_opt_size, primer_internal_max_poly_x, primer_max_poly_x, primer_gc_clamp, primer_max_end_gc\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    primers = [\"ATGCGG\", \"CTAGGG\"]\n",
    "    amplified_sequence = \"ATGCGGCTAGGG\"\n",
    "\n",
    "    primer_parameters = get_primer_parameters(primers, amplified_sequence)\n",
    "\n",
    "    print(\"PRIMER_INTERNAL_OPT_SIZE:\", primer_parameters[0])\n",
    "    print(\"PRIMER_INTERNAL_MAX_POLY_X:\", primer_parameters[1])\n",
    "    print(\"PRIMER_MAX_POLY_X:\", primer_parameters[2])\n",
    "    print(\"PRIMER_GC_CLAMP:\", primer_parameters[3])\n",
    "    print(\"PRIMER_MAX_END_GC:\", primer_parameters[4])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# getting primer3 output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6510, 6528)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_sequence_location('AGAGTTGGTGCGGCGTAA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "considered 634, unacceptable product size 31, ok 603\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556 14.880899385848068\n",
      "(6510, 6528)\n"
     ]
    }
   ],
   "source": [
    "sequence = extract_sequence_from_fasta(6511-5, 6959, padding = 0, fasta_file= '/mnt/storage10/lwang/Projects/Amplicone_design_tool/MTB-h37rv_asm19595v2-eg18.fa', sequence_id='Chromosome')\n",
    "\n",
    "middle = extract_sequence_from_fasta(30000, 30500, padding = 0, fasta_file= '/mnt/storage10/lwang/Projects/Amplicone_design_tool/GCF_000002765.5_GCA_000002765_genomic.fna', sequence_id='NC_004325.2')\n",
    "\n",
    "# sequence[:250]+'N'*100+sequence[-250:]\n",
    "# print(sequence)\n",
    "results = bindings.design_primers(\n",
    "            seq_args={\n",
    "                'SEQUENCE_ID': 'primer',\n",
    "\n",
    "                'SEQUENCE_TEMPLATE': sequence,\n",
    "                'SEQUENCE_INCLUDED_REGION': [0,len(sequence)],\n",
    "                # 'SEQUENCE_INCLUDED_REGION': [(250,len(sequence)-250)],\n",
    "                # 'SEQUENCE_EXCLUDED_REGION': [(200,300)],\n",
    "                \n",
    "                # 'SEQUENCE_TEMPLATE': sequence[:250]+middle+sequence[-250:],\n",
    "                # 'SEQUENCE_INCLUDED_REGION': [250,750],\n",
    "                # 'SEQUENCE_EXCLUDED_REGION': [250,750],\n",
    "\n",
    "                # 'SEQUENCE_INCLUDED_REGION': sequence[:251],\n",
    "            },\n",
    "            global_args={\n",
    "                'PRIMER_NUM_RETURN': 600,\n",
    "                # 'PRIMER_OPT_SIZE': 20,\n",
    "                'PRIMER_INTERNAL_OPT_SIZE': 20,\n",
    "                'PRIMER_INTERNAL_MAX_POLY_X': 0,\n",
    "                'PRIMER_MAX_POLY_X': 0,\n",
    "                # 'PRIMER_GC_CLAMP': 0.6666666666666666,\n",
    "                # 'PRIMER_MAX_END_GC': 1.0,\n",
    "\n",
    "                'PRIMER_PICK_INTERNAL_OLIGO': 0,\n",
    "                'PRIMER_INTERNAL_MAX_SELF_END': 8,\n",
    "                # 'PRIMER_MIN_THREE_PRIME_DISTANCE':10,\n",
    "                # 'PRIMER_MIN_FIVE_PRIME_DISTANCE':10,\n",
    "                'PRIMER_MIN_SIZE': 16,\n",
    "                'PRIMER_MAX_SIZE': 25,\n",
    "                'PRIMER_OPT_TM': 50.0,\n",
    "                'PRIMER_MIN_TM': 47.0,\n",
    "                'PRIMER_MAX_TM': 69,#59.0,\n",
    "                'PRIMER_MIN_GC': 45,#45.0,\n",
    "                'PRIMER_MAX_GC': 68,#79.0,\n",
    "                'PRIMER_MAX_POLY_X': 5,\n",
    "                'PRIMER_INTERNAL_MAX_POLY_X': 5,\n",
    "                'PRIMER_SALT_MONOVALENT': 50.0,\n",
    "                'PRIMER_DNA_CONC': 500.0,\n",
    "                'PRIMER_MAX_NS_ACCEPTED': 2,\n",
    "                'PRIMER_MAX_SELF_ANY': 5,\n",
    "                'PRIMER_MAX_SELF_END': 2,\n",
    "                'PRIMER_PAIR_MAX_COMPL_ANY': 5,\n",
    "                'PRIMER_PAIR_MAX_COMPL_END': 2,\n",
    "                'PRIMER_PRODUCT_SIZE_RANGE': '440-470',\n",
    "                # 'PRIMER_PRODUCT_SIZE_RANGE': '550-650',\n",
    "                # 'PRIMER_PRODUCT_SIZE_RANGE': [\n",
    "                #     # [950,1050]\n",
    "                #     [len(sequence)-350,len(sequence)-250]\n",
    "                # ],\n",
    "            })\n",
    "\n",
    "print(results['PRIMER_PAIR_EXPLAIN'])\n",
    "\n",
    "# checking primer similarity\n",
    "query_seq = 'AGAGTTGGTGCGGCGTAA'\n",
    "threshold = 0.99\n",
    "primer_loc = [6510, 6928]\n",
    "for x in results['PRIMER_LEFT']:\n",
    "    # print(x['SEQUENCE'], x['TM'], x['GC_PERCENT'])\n",
    "    seq_loc = find_sequence_location(x['SEQUENCE'])\n",
    "    # if check_similarity(x['SEQUENCE'], query_seq, threshold=threshold) & seq_loc[0]>primer_loc[0]-10 & seq_loc[1]<primer_loc[1]+10:\n",
    "    if check_similarity(x['SEQUENCE'], query_seq, threshold=threshold):\n",
    "        # checking to see if the primer is within 10 bp of the known primer and have a similarity of that passses the threshold\n",
    "        # print(check_similarity(x['SEQUENCE'], query_seq, threshold=threshold))\n",
    "        print(x['SEQUENCE'], x['TM'], x['GC_PERCENT'], x['PENALTY'])\n",
    "        print(seq_loc)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug</th>\n",
       "      <th>primer name</th>\n",
       "      <th>forward</th>\n",
       "      <th>reverse</th>\n",
       "      <th>Amplicon size</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Fluoroquinolones</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>AGAGTTGGTGCGGCGTAA</td>\n",
       "      <td>GCCACTTGAGTTTGTACAGC</td>\n",
       "      <td>449.0</td>\n",
       "      <td>6511.0</td>\n",
       "      <td>6959.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fluoroquinolones</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>GTTGACATCGAGCAGGAGAT</td>\n",
       "      <td>AAATCGACTGTCTCCTCGTC</td>\n",
       "      <td>404.0</td>\n",
       "      <td>7353.0</td>\n",
       "      <td>7756.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rifampicin</td>\n",
       "      <td>rpoB_1</td>\n",
       "      <td>GGTGGCCTGGAAGAGGT</td>\n",
       "      <td>AGAACCCGAACCGCTCG</td>\n",
       "      <td>514.0</td>\n",
       "      <td>760038.0</td>\n",
       "      <td>760551.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rifampicin</td>\n",
       "      <td>rpoB_2</td>\n",
       "      <td>CCCCGACCAAAGAGTCA</td>\n",
       "      <td>CTCACGTGACAGACCGC</td>\n",
       "      <td>542.0</td>\n",
       "      <td>760645.0</td>\n",
       "      <td>761186.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rifampicin</td>\n",
       "      <td>rifampicin</td>\n",
       "      <td>TGCTCACCTCGATCCACATC</td>\n",
       "      <td>ATCTTGCGTTTCTGAGCCAC</td>\n",
       "      <td>400.0</td>\n",
       "      <td>762025.0</td>\n",
       "      <td>762424.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Rifampicin</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>TGCGCCCGATGGTGCAGCT</td>\n",
       "      <td>CGGCGATGACCTCTTCG</td>\n",
       "      <td>507.0</td>\n",
       "      <td>764340.0</td>\n",
       "      <td>764846.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Bedaquiline/Clofazimine</td>\n",
       "      <td>mmpR5</td>\n",
       "      <td>TTCGAGTCCAGGAGTTTGAC</td>\n",
       "      <td>GCTCATCAGTCGTCCTCTC</td>\n",
       "      <td>425.0</td>\n",
       "      <td>779068.0</td>\n",
       "      <td>779492.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Streptomycin</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>GTGAAAGCGCCCAAGATAGA</td>\n",
       "      <td>CATCAGCCCTTCTCCTTCTT</td>\n",
       "      <td>407.0</td>\n",
       "      <td>781530.0</td>\n",
       "      <td>781936.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Streptomycin**</td>\n",
       "      <td>rrs_1</td>\n",
       "      <td>ATTGCACAATGGGCGCAAGC</td>\n",
       "      <td>CAAGGAAGGAAACCCACACC</td>\n",
       "      <td>475.0</td>\n",
       "      <td>1472209.0</td>\n",
       "      <td>1472684.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Streptomycin**</td>\n",
       "      <td>rrs_3</td>\n",
       "      <td>CACAGGACGCGTCTAGAGAT</td>\n",
       "      <td>AGGTGATCCAGCCGCACCTT</td>\n",
       "      <td>539.0</td>\n",
       "      <td>1472837.0</td>\n",
       "      <td>1473375.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Isoniazid* (Rv1482c-fabG1)</td>\n",
       "      <td>fabG1</td>\n",
       "      <td>CCAGAAAGGGATCCGTCATG</td>\n",
       "      <td>GACCCTGGTGCTCTTCTAC</td>\n",
       "      <td>400.0</td>\n",
       "      <td>1673281.0</td>\n",
       "      <td>1673680.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Isoniazid*</td>\n",
       "      <td>inhA_1</td>\n",
       "      <td>CAAACGGATTCTGGTTAGCG</td>\n",
       "      <td>GACCGTCATCCAGTTGTAGG</td>\n",
       "      <td>469.0</td>\n",
       "      <td>1674222.0</td>\n",
       "      <td>1674690.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Isoniazid*</td>\n",
       "      <td>inhA_2</td>\n",
       "      <td>CTGCTGCCGATCATGAACC</td>\n",
       "      <td>GCGTTCTAGAGCAATTGGGT</td>\n",
       "      <td>416.0</td>\n",
       "      <td>1674601.0</td>\n",
       "      <td>1675016.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PAS</td>\n",
       "      <td>tlyA</td>\n",
       "      <td>TTGTTGGCCGCCCTGGA</td>\n",
       "      <td>CCAACGAGATGAACGACAGGTCGG</td>\n",
       "      <td>524.0</td>\n",
       "      <td>1917897.0</td>\n",
       "      <td>1918420.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Isoniazid</td>\n",
       "      <td>katG_1</td>\n",
       "      <td>GCCGATCAACCCGAATCA</td>\n",
       "      <td>AGTACATGCTGCTCGACAAG</td>\n",
       "      <td>419.0</td>\n",
       "      <td>2153874.0</td>\n",
       "      <td>2154292.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Isoniazid</td>\n",
       "      <td>katG_2</td>\n",
       "      <td>ATTGCCAAGCGCCAGCA</td>\n",
       "      <td>TCAAGACGTTCGGGTTCGG</td>\n",
       "      <td>525.0</td>\n",
       "      <td>2155055.0</td>\n",
       "      <td>2155579.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Pyrazinamide</td>\n",
       "      <td>pncA_1</td>\n",
       "      <td>CCAACAGTTCATCCCGGTT</td>\n",
       "      <td>GACTATTCCTCGTCGTGGC</td>\n",
       "      <td>405.0</td>\n",
       "      <td>2288651.0</td>\n",
       "      <td>2289055.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Pyrazinamide</td>\n",
       "      <td>pncA_2</td>\n",
       "      <td>AGGCACCCTTGTAGAACAC</td>\n",
       "      <td>TTGTCGCTCACTACATCACC</td>\n",
       "      <td>457.0</td>\n",
       "      <td>2288947.0</td>\n",
       "      <td>2289403.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Isoniazid (oxyR'-ahpC)</td>\n",
       "      <td>ahpC</td>\n",
       "      <td>GCCGGCTAGCACCTCTT</td>\n",
       "      <td>GCGCTTGATGTCGGAGA</td>\n",
       "      <td>479.0</td>\n",
       "      <td>2726062.0</td>\n",
       "      <td>2726540.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PAS</td>\n",
       "      <td>folC</td>\n",
       "      <td>GACCACTGCCACGTCGA</td>\n",
       "      <td>TTCGACGAATTCCGGCC</td>\n",
       "      <td>514.0</td>\n",
       "      <td>2747080.0</td>\n",
       "      <td>2747593.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PAS</td>\n",
       "      <td>thyX</td>\n",
       "      <td>GGTCTTGGGATTGGGCTTGG</td>\n",
       "      <td>GCCGCTGGACACTTGCCTAT</td>\n",
       "      <td>460.0</td>\n",
       "      <td>3067784.0</td>\n",
       "      <td>3068244.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PAS</td>\n",
       "      <td>thyA</td>\n",
       "      <td>CATACCGCGACTGGAGCTTT</td>\n",
       "      <td>ATCATCGTGTCGGCCTGGAA</td>\n",
       "      <td>409.0</td>\n",
       "      <td>3073681.0</td>\n",
       "      <td>3074090.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Ethambutol</td>\n",
       "      <td>embA</td>\n",
       "      <td>AGCTACCTCAAAGACGACTG</td>\n",
       "      <td>AGCAGGGGATCGAGATGT</td>\n",
       "      <td>461.0</td>\n",
       "      <td>4243001.0</td>\n",
       "      <td>4243461.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Ethambutol</td>\n",
       "      <td>embB</td>\n",
       "      <td>GCACCTTCACCCTGACCG</td>\n",
       "      <td>GGCTGCACACCCAGTGTG</td>\n",
       "      <td>517.0</td>\n",
       "      <td>4247334.0</td>\n",
       "      <td>4247850.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Ethambutol</td>\n",
       "      <td>embB_2</td>\n",
       "      <td>GATGCTGCGGATCTTGGTG</td>\n",
       "      <td>GATACCAACACCGTCGTCAG</td>\n",
       "      <td>461.0</td>\n",
       "      <td>4247896.0</td>\n",
       "      <td>4248356.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Ethionamide</td>\n",
       "      <td>ethA_1</td>\n",
       "      <td>TCGCTAAAGCTAAACCCC</td>\n",
       "      <td>TGAACGGTTCACCGCGA</td>\n",
       "      <td>529.0</td>\n",
       "      <td>4325995.0</td>\n",
       "      <td>4326523.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Ethionamide</td>\n",
       "      <td>ethA_2</td>\n",
       "      <td>CCGGATGAGCCGGATGT</td>\n",
       "      <td>CCATTCGTCACGGGAAGG</td>\n",
       "      <td>497.0</td>\n",
       "      <td>4326067.0</td>\n",
       "      <td>4326563.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Streptomycin</td>\n",
       "      <td>gid</td>\n",
       "      <td>CCTTTGATGGCGAGCATTC</td>\n",
       "      <td>AGTGCGTAATGTCTCCGATC</td>\n",
       "      <td>499.0</td>\n",
       "      <td>4407712.0</td>\n",
       "      <td>4408210.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Aminoglycosides (eis-Rv2417c)</td>\n",
       "      <td>eis</td>\n",
       "      <td>AGGGTCACAGTCACAGAATC</td>\n",
       "      <td>AAACTTGTTCTGGTCCAACG</td>\n",
       "      <td>429.0</td>\n",
       "      <td>2715319.0</td>\n",
       "      <td>2715747.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>D-Cycloserine</td>\n",
       "      <td>ald</td>\n",
       "      <td>TCGCGGAACTAACCCGT</td>\n",
       "      <td>ATGCGGGCTGCGTTGTA</td>\n",
       "      <td>490.0</td>\n",
       "      <td>3086887.0</td>\n",
       "      <td>3087376.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>D-Cycloserine</td>\n",
       "      <td>alr</td>\n",
       "      <td>ATGGCTTGGCGTAACGC</td>\n",
       "      <td>CGTTGGCCATGACACCG</td>\n",
       "      <td>484.0</td>\n",
       "      <td>3840873.0</td>\n",
       "      <td>3841356.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Ethambutol</td>\n",
       "      <td>embC</td>\n",
       "      <td>AGATCGTCGGGGTGTTCA</td>\n",
       "      <td>GTCGTAGTACCAGCCGAAAG</td>\n",
       "      <td>401.0</td>\n",
       "      <td>4240449.0</td>\n",
       "      <td>4240849.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Ethionamide</td>\n",
       "      <td>ethA_3</td>\n",
       "      <td>CGGAGTCGGAGCGAATT</td>\n",
       "      <td>GTCGGCTTGATTGACCACC</td>\n",
       "      <td>499.0</td>\n",
       "      <td>4327302.0</td>\n",
       "      <td>4327800.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Ethionamide</td>\n",
       "      <td>ethR</td>\n",
       "      <td>GGCCGACGTTCTACTTCTAT</td>\n",
       "      <td>GTTCTCGCCGTAAATGCTG</td>\n",
       "      <td>473.0</td>\n",
       "      <td>4327721.0</td>\n",
       "      <td>4328193.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Ethambutol</td>\n",
       "      <td>embB_3</td>\n",
       "      <td>CAGACTGTGGTGCTCGAATAC</td>\n",
       "      <td>AGCAACAGGTCGGTGATCC</td>\n",
       "      <td>446.0</td>\n",
       "      <td>4249199.0</td>\n",
       "      <td>4249644.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Rifampicin</td>\n",
       "      <td>rpoB (1_2)</td>\n",
       "      <td>CTGTCTCCGATCGAGGACTTCT</td>\n",
       "      <td>AGACGAGGGCACGTACTCCA</td>\n",
       "      <td>1446.0</td>\n",
       "      <td>760065.0</td>\n",
       "      <td>761510.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Isoniazid</td>\n",
       "      <td>katG (1_2)</td>\n",
       "      <td>CAGGGCCGATCAACCCGAAT</td>\n",
       "      <td>CGGTGCAGATGGGGCTGATC</td>\n",
       "      <td>1578.0</td>\n",
       "      <td>2153870.0</td>\n",
       "      <td>2155447.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Streptomycin</td>\n",
       "      <td>rrs_2</td>\n",
       "      <td>ATTGCACAATGGGCGCAAGC</td>\n",
       "      <td>AGGTGATCCAGCCGCACCTT</td>\n",
       "      <td>1166.0</td>\n",
       "      <td>1472210.0</td>\n",
       "      <td>1473375.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Isoniazid</td>\n",
       "      <td>katG (1_2)</td>\n",
       "      <td>CAGGGCCGATCAACCCGAAT</td>\n",
       "      <td>CGGTGCAGATGGGGCTGATC</td>\n",
       "      <td>1578.0</td>\n",
       "      <td>2153870.0</td>\n",
       "      <td>2155447.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Drug primer name                 forward  \\\n",
       "0                Fluoroquinolones        gyrB      AGAGTTGGTGCGGCGTAA   \n",
       "1                Fluoroquinolones        gyrA    GTTGACATCGAGCAGGAGAT   \n",
       "2                      Rifampicin      rpoB_1       GGTGGCCTGGAAGAGGT   \n",
       "3                      Rifampicin      rpoB_2       CCCCGACCAAAGAGTCA   \n",
       "4                      Rifampicin  rifampicin    TGCTCACCTCGATCCACATC   \n",
       "5                      Rifampicin        rpoC     TGCGCCCGATGGTGCAGCT   \n",
       "6         Bedaquiline/Clofazimine       mmpR5    TTCGAGTCCAGGAGTTTGAC   \n",
       "7                    Streptomycin        rpsL    GTGAAAGCGCCCAAGATAGA   \n",
       "8                  Streptomycin**       rrs_1    ATTGCACAATGGGCGCAAGC   \n",
       "9                  Streptomycin**       rrs_3    CACAGGACGCGTCTAGAGAT   \n",
       "10     Isoniazid* (Rv1482c-fabG1)       fabG1    CCAGAAAGGGATCCGTCATG   \n",
       "11                     Isoniazid*      inhA_1    CAAACGGATTCTGGTTAGCG   \n",
       "12                     Isoniazid*      inhA_2     CTGCTGCCGATCATGAACC   \n",
       "13                            PAS        tlyA       TTGTTGGCCGCCCTGGA   \n",
       "14                      Isoniazid      katG_1      GCCGATCAACCCGAATCA   \n",
       "15                      Isoniazid      katG_2       ATTGCCAAGCGCCAGCA   \n",
       "16                   Pyrazinamide      pncA_1     CCAACAGTTCATCCCGGTT   \n",
       "17                   Pyrazinamide      pncA_2     AGGCACCCTTGTAGAACAC   \n",
       "18         Isoniazid (oxyR'-ahpC)        ahpC       GCCGGCTAGCACCTCTT   \n",
       "19                            PAS        folC       GACCACTGCCACGTCGA   \n",
       "20                            PAS        thyX    GGTCTTGGGATTGGGCTTGG   \n",
       "21                            PAS        thyA    CATACCGCGACTGGAGCTTT   \n",
       "22                     Ethambutol        embA    AGCTACCTCAAAGACGACTG   \n",
       "23                     Ethambutol        embB      GCACCTTCACCCTGACCG   \n",
       "24                     Ethambutol      embB_2     GATGCTGCGGATCTTGGTG   \n",
       "25                    Ethionamide      ethA_1      TCGCTAAAGCTAAACCCC   \n",
       "26                    Ethionamide      ethA_2       CCGGATGAGCCGGATGT   \n",
       "27                   Streptomycin         gid     CCTTTGATGGCGAGCATTC   \n",
       "28  Aminoglycosides (eis-Rv2417c)         eis    AGGGTCACAGTCACAGAATC   \n",
       "29                  D-Cycloserine         ald       TCGCGGAACTAACCCGT   \n",
       "30                  D-Cycloserine         alr       ATGGCTTGGCGTAACGC   \n",
       "31                     Ethambutol        embC      AGATCGTCGGGGTGTTCA   \n",
       "32                    Ethionamide      ethA_3       CGGAGTCGGAGCGAATT   \n",
       "33                    Ethionamide        ethR    GGCCGACGTTCTACTTCTAT   \n",
       "34                     Ethambutol      embB_3   CAGACTGTGGTGCTCGAATAC   \n",
       "35                     Rifampicin  rpoB (1_2)  CTGTCTCCGATCGAGGACTTCT   \n",
       "36                      Isoniazid  katG (1_2)    CAGGGCCGATCAACCCGAAT   \n",
       "37                   Streptomycin       rrs_2    ATTGCACAATGGGCGCAAGC   \n",
       "38                      Isoniazid  katG (1_2)    CAGGGCCGATCAACCCGAAT   \n",
       "\n",
       "                     reverse  Amplicon size      Start        End  \n",
       "0       GCCACTTGAGTTTGTACAGC          449.0     6511.0     6959.0  \n",
       "1       AAATCGACTGTCTCCTCGTC          404.0     7353.0     7756.0  \n",
       "2          AGAACCCGAACCGCTCG          514.0   760038.0   760551.0  \n",
       "3          CTCACGTGACAGACCGC          542.0   760645.0   761186.0  \n",
       "4       ATCTTGCGTTTCTGAGCCAC          400.0   762025.0   762424.0  \n",
       "5          CGGCGATGACCTCTTCG          507.0   764340.0   764846.0  \n",
       "6        GCTCATCAGTCGTCCTCTC          425.0   779068.0   779492.0  \n",
       "7       CATCAGCCCTTCTCCTTCTT          407.0   781530.0   781936.0  \n",
       "8       CAAGGAAGGAAACCCACACC          475.0  1472209.0  1472684.0  \n",
       "9       AGGTGATCCAGCCGCACCTT          539.0  1472837.0  1473375.0  \n",
       "10       GACCCTGGTGCTCTTCTAC          400.0  1673281.0  1673680.0  \n",
       "11      GACCGTCATCCAGTTGTAGG          469.0  1674222.0  1674690.0  \n",
       "12      GCGTTCTAGAGCAATTGGGT          416.0  1674601.0  1675016.0  \n",
       "13  CCAACGAGATGAACGACAGGTCGG          524.0  1917897.0  1918420.0  \n",
       "14      AGTACATGCTGCTCGACAAG          419.0  2153874.0  2154292.0  \n",
       "15       TCAAGACGTTCGGGTTCGG          525.0  2155055.0  2155579.0  \n",
       "16       GACTATTCCTCGTCGTGGC          405.0  2288651.0  2289055.0  \n",
       "17      TTGTCGCTCACTACATCACC          457.0  2288947.0  2289403.0  \n",
       "18         GCGCTTGATGTCGGAGA          479.0  2726062.0  2726540.0  \n",
       "19         TTCGACGAATTCCGGCC          514.0  2747080.0  2747593.0  \n",
       "20      GCCGCTGGACACTTGCCTAT          460.0  3067784.0  3068244.0  \n",
       "21      ATCATCGTGTCGGCCTGGAA          409.0  3073681.0  3074090.0  \n",
       "22        AGCAGGGGATCGAGATGT          461.0  4243001.0  4243461.0  \n",
       "23        GGCTGCACACCCAGTGTG          517.0  4247334.0  4247850.0  \n",
       "24      GATACCAACACCGTCGTCAG          461.0  4247896.0  4248356.0  \n",
       "25         TGAACGGTTCACCGCGA          529.0  4325995.0  4326523.0  \n",
       "26        CCATTCGTCACGGGAAGG          497.0  4326067.0  4326563.0  \n",
       "27      AGTGCGTAATGTCTCCGATC          499.0  4407712.0  4408210.0  \n",
       "28      AAACTTGTTCTGGTCCAACG          429.0  2715319.0  2715747.0  \n",
       "29         ATGCGGGCTGCGTTGTA          490.0  3086887.0  3087376.0  \n",
       "30         CGTTGGCCATGACACCG          484.0  3840873.0  3841356.0  \n",
       "31      GTCGTAGTACCAGCCGAAAG          401.0  4240449.0  4240849.0  \n",
       "32       GTCGGCTTGATTGACCACC          499.0  4327302.0  4327800.0  \n",
       "33       GTTCTCGCCGTAAATGCTG          473.0  4327721.0  4328193.0  \n",
       "34       AGCAACAGGTCGGTGATCC          446.0  4249199.0  4249644.0  \n",
       "35      AGACGAGGGCACGTACTCCA         1446.0   760065.0   761510.0  \n",
       "36      CGGTGCAGATGGGGCTGATC         1578.0  2153870.0  2155447.0  \n",
       "37      AGGTGATCCAGCCGCACCTT         1166.0  1472210.0  1473375.0  \n",
       "38      CGGTGCAGATGGGGCTGATC         1578.0  2153870.0  2155447.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "known_primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CGGGCAGGGCCGATCAACCCGAATCAGCGCACGTCGAACCTGTCGAGGTTCATCACCTTGTCCCAGGCAGCGACGAAGTCCTGCACGAACTTCGGCTGCGCGTCATCGGCGCCATAGACCTCGACAAGCGCCCGCAACTCCGAGTTGGACCCGAAGACCAGGTCCACGCGGCTGCCGGTCCACTTCACCTTGCCACTGCCATCCTTGCCCTGGTAGGTCCCGTCATCTGCTGGCGAGGGCTCCCAGGTGATACCCATGTCGAGCAGGTTCACGAAGAAGTCGTTGGTCAGTGACTCGGAGGCCTCGGTGAACACGCCCAGCGGTAAGCGCTTGTAGTTTGCGCCGAGGACGCGCAGGCCACCTACCAGCACCGTCATCTCAGGGGCACTGAGCGTAAGCAGGTTCGCCTTGTCGAGCAGCATGTACTCGGCCGGCAACGGGTTGCCCTTTCCGAGGTAGTTTCGGAAGCCATCTGCCTTGGGCTCCAGCACGGCAAAGGATTCCACGTCGGTTTGTTCCTGCGACGCATCCGTGCGGCCCGGGGTGAAGGGCACCGTGATGTTGTGGCCAGCCGCCTTTGCTGCTTTCTCTATGGCGGCACAGCCACCGAGCACGACGAGGTCGGCGAAGGACACTTTGATGTTCCCCGGCGCCGCGGAGTTGAATGACTCCTGGATCTCTTCCAGGGTGCGAATGACCTTGCGCAGATCCCCGTCGGGGTCGTTGACCTCCCACCCGACTTGTGGCTGCAGGCGGATGCGACCACCGTTGGCGCCGCCGCGCTTGTCGCTACCACGGAACGACGACGCCGCCGCCCATGCGGTCGAAACTAGCTGTGAGACAGTCAATCCCGATGCCCGGATCTGGCTCTTAAGGCTGGCAATCTCGGCTTCGCCGACGAGGTCGTGGCTGACCGCAGGGACCGGATCCTGCCACAGCAGGGTCTGCTTGGGGACCAGCGGCCCAAGGTATCTCGCAACGGGACCCATGTCTCGGTGGATCAGCTTGTACCAGGCCTTGGCGAACTCGTCGGCCAATTCCTCGGGGTGTTCCAGCCAGCGACGCGTGATCCGCTCATAGATCGGATCCACCCGCAGCGAGAGGTCAGTGGCCAGCATCGTCGGGGAGCGCCCTGGCCCGCCGAACGGGTCCGGGATGGTGCCGGCACCGGCGCCGTCCTTGGCGGTGTATTGCCAAGCGCCAGCAGGGCTCTTCGTCAGCTCCCACTCGTAGCCGTACAGGATCTCGAGGAAACTGTTGTCCCATTTCGTCGGGGTGTTCGTCCATACGACCTCGATGCCGCTGGTGATCGCGTCCTTACCGGTTCCGGTGCCATACGAGCTCTTCCAGCCCAAGCCCATCTGCTCCAGCGGAGCAGCCTCGGGTTCGGGGCCGACCAGATCGGCCGGGCCGGCGCCATGGGTCTTACCGAAAGTGTGACCGCCGACGATCAGCGCCGCTGTTTCGACGTCGTTCATGGCCATGCGCCGAAACGTCTCGCGAATGTCGACCGCCGCGGCCATGGGGTCCGGGTTGCCGTTCGGCCCCTCCGGGTTCACGTAGATCAGCCCCATCTGCACCG'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "PRIMER_MAX_END_GC must be between 0 to 5",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[21], line 11\u001b[0m\n\u001b[1;32m      8\u001b[0m     lower_range \u001b[39m=\u001b[39m \u001b[39mint\u001b[39m(row[\u001b[39m'\u001b[39m\u001b[39mAmplicon size\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m-\u001b[39m \u001b[39m10\u001b[39m)\n\u001b[1;32m      9\u001b[0m     \u001b[39m# sequence[:250]+'N'*100+sequence[-250:]\u001b[39;00m\n\u001b[1;32m     10\u001b[0m     \u001b[39m# print(sequence)\u001b[39;00m\n\u001b[0;32m---> 11\u001b[0m     results \u001b[39m=\u001b[39m bindings\u001b[39m.\u001b[39;49mdesign_primers(\n\u001b[1;32m     12\u001b[0m                 seq_args\u001b[39m=\u001b[39;49m{\n\u001b[1;32m     13\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mSEQUENCE_ID\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m'\u001b[39;49m\u001b[39mprimer\u001b[39;49m\u001b[39m'\u001b[39;49m,\n\u001b[1;32m     14\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mSEQUENCE_TEMPLATE\u001b[39;49m\u001b[39m'\u001b[39;49m: sequence,\n\u001b[1;32m     15\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mSEQUENCE_INCLUDED_REGION\u001b[39;49m\u001b[39m'\u001b[39;49m: [\u001b[39m0\u001b[39;49m,\u001b[39mlen\u001b[39;49m(sequence)],\n\u001b[1;32m     16\u001b[0m                     \u001b[39m# 'SEQUENCE_INCLUDED_REGION': [(250,len(sequence)-250)],\u001b[39;49;00m\n\u001b[1;32m     17\u001b[0m                     \u001b[39m# 'SEQUENCE_EXCLUDED_REGION': [(200,300)],\u001b[39;49;00m\n\u001b[1;32m     18\u001b[0m                     \n\u001b[1;32m     19\u001b[0m                     \u001b[39m# 'SEQUENCE_TEMPLATE': sequence[:250]+middle+sequence[-250:],\u001b[39;49;00m\n\u001b[1;32m     20\u001b[0m                     \u001b[39m# 'SEQUENCE_INCLUDED_REGION': [250,750],\u001b[39;49;00m\n\u001b[1;32m     21\u001b[0m                     \u001b[39m# 'SEQUENCE_EXCLUDED_REGION': [250,750],\u001b[39;49;00m\n\u001b[1;32m     22\u001b[0m \n\u001b[1;32m     23\u001b[0m                     \u001b[39m# 'SEQUENCE_INCLUDED_REGION': sequence[:251],\u001b[39;49;00m\n\u001b[1;32m     24\u001b[0m                 \n\u001b[1;32m     25\u001b[0m                 },\n\u001b[1;32m     26\u001b[0m                 global_args\u001b[39m=\u001b[39;49m{\n\u001b[1;32m     27\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MAX_END_GC\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m10\u001b[39;49m,\n\u001b[1;32m     28\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_PICK_ANYWAY\u001b[39;49m\u001b[39m'\u001b[39;49m:\u001b[39m1\u001b[39;49m,\n\u001b[1;32m     29\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_NUM_RETURN\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m600\u001b[39;49m,\n\u001b[1;32m     30\u001b[0m                     \u001b[39m# 'PRIMER_OPT_SIZE': 20,\u001b[39;49;00m\n\u001b[1;32m     31\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_PICK_INTERNAL_OLIGO\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m1\u001b[39;49m,\n\u001b[1;32m     32\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_INTERNAL_MAX_SELF_END\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m8\u001b[39;49m,\n\u001b[1;32m     33\u001b[0m                     \u001b[39m# 'PRIMER_MIN_THREE_PRIME_DISTANCE':10,\u001b[39;49;00m\n\u001b[1;32m     34\u001b[0m                     \u001b[39m# 'PRIMER_MIN_FIVE_PRIME_DISTANCE':10,\u001b[39;49;00m\n\u001b[1;32m     35\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MIN_SIZE\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m16\u001b[39;49m,\n\u001b[1;32m     36\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MAX_SIZE\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m25\u001b[39;49m,\n\u001b[1;32m     37\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_OPT_TM\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m50.0\u001b[39;49m,\n\u001b[1;32m     38\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MIN_TM\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m45\u001b[39;49m,\n\u001b[1;32m     39\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MAX_TM\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m71\u001b[39;49m,\u001b[39m#59.0,\u001b[39;49;00m\n\u001b[1;32m     40\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MIN_GC\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m45\u001b[39;49m,\u001b[39m#45.0,\u001b[39;49;00m\n\u001b[1;32m     41\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MAX_GC\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m68\u001b[39;49m,\u001b[39m#79.0,\u001b[39;49;00m\n\u001b[1;32m     42\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MAX_POLY_X\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m5\u001b[39;49m,\n\u001b[1;32m     43\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_INTERNAL_MAX_POLY_X\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m5\u001b[39;49m,\n\u001b[1;32m     44\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_SALT_MONOVALENT\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m50.0\u001b[39;49m,\n\u001b[1;32m     45\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_DNA_CONC\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m50.0\u001b[39;49m,\n\u001b[1;32m     46\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MAX_NS_ACCEPTED\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m0\u001b[39;49m,\n\u001b[1;32m     47\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MAX_SELF_ANY\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m12\u001b[39;49m,\n\u001b[1;32m     48\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_MAX_SELF_END\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m8\u001b[39;49m,\n\u001b[1;32m     49\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_PAIR_MAX_COMPL_ANY\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m12\u001b[39;49m,\n\u001b[1;32m     50\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_PAIR_MAX_COMPL_END\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39m8\u001b[39;49m,\n\u001b[1;32m     51\u001b[0m                     \u001b[39m'\u001b[39;49m\u001b[39mPRIMER_PRODUCT_SIZE_RANGE\u001b[39;49m\u001b[39m'\u001b[39;49m: \u001b[39mf\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m{\u001b[39;49;00mlower_range\u001b[39m}\u001b[39;49;00m\u001b[39m-\u001b[39;49m\u001b[39m{\u001b[39;49;00mupper_range\u001b[39m}\u001b[39;49;00m\u001b[39m'\u001b[39;49m,\n\u001b[1;32m     52\u001b[0m                     \u001b[39m# 'PRIMER_PRODUCT_SIZE_RANGE': '550-650',\u001b[39;49;00m\n\u001b[1;32m     53\u001b[0m                     \u001b[39m# 'PRIMER_PRODUCT_SIZE_RANGE': [\u001b[39;49;00m\n\u001b[1;32m     54\u001b[0m                     \u001b[39m#     # [950,1050]\u001b[39;49;00m\n\u001b[1;32m     55\u001b[0m                     \u001b[39m#     [len(sequence)-350,len(sequence)-250]\u001b[39;49;00m\n\u001b[1;32m     56\u001b[0m                     \u001b[39m# ],\u001b[39;49;00m\n\u001b[1;32m     57\u001b[0m                 })\n\u001b[1;32m     59\u001b[0m \u001b[39m# PRIMER_MIN_TM 45.0\u001b[39;00m\n\u001b[1;32m     60\u001b[0m \u001b[39m# PRIMER_MAX_TM 71\u001b[39;00m\n\u001b[1;32m     61\u001b[0m \u001b[39m# PRIMER_MIN_GC 47\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m     \u001b[39m# checking primer similarity\u001b[39;00m\n\u001b[1;32m     66\u001b[0m     query_seq \u001b[39m=\u001b[39m row[\u001b[39m'\u001b[39m\u001b[39mforward\u001b[39m\u001b[39m'\u001b[39m]\n",
      "File \u001b[0;32m~/miniconda3/envs/ml-s10/lib/python3.11/site-packages/primer3/bindings.py:550\u001b[0m, in \u001b[0;36mdesign_primers\u001b[0;34m(seq_args, global_args, misprime_lib, mishyb_lib)\u001b[0m\n\u001b[1;32m    528\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mdesign_primers\u001b[39m(\n\u001b[1;32m    529\u001b[0m         seq_args: Dict[\u001b[39mstr\u001b[39m, Any],\n\u001b[1;32m    530\u001b[0m         global_args: Dict[\u001b[39mstr\u001b[39m, Any],\n\u001b[1;32m    531\u001b[0m         misprime_lib: Optional[Dict[\u001b[39mstr\u001b[39m, Any]] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    532\u001b[0m         mishyb_lib: Optional[Dict[\u001b[39mstr\u001b[39m, Any]] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    533\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Dict[\u001b[39mstr\u001b[39m, Any]:\n\u001b[1;32m    534\u001b[0m \u001b[39m    \u001b[39m\u001b[39m'''Run the Primer3 design process.\u001b[39;00m\n\u001b[1;32m    535\u001b[0m \u001b[39m    This is a wrapper around :meth:`_ThermoAnalysis.run_design`.\u001b[39;00m\n\u001b[1;32m    536\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    548\u001b[0m \n\u001b[1;32m    549\u001b[0m \u001b[39m    '''\u001b[39;00m\n\u001b[0;32m--> 550\u001b[0m     \u001b[39mreturn\u001b[39;00m THERMO_ANALYSIS\u001b[39m.\u001b[39;49mrun_design(\n\u001b[1;32m    551\u001b[0m         global_args\u001b[39m=\u001b[39;49mglobal_args,\n\u001b[1;32m    552\u001b[0m         seq_args\u001b[39m=\u001b[39;49mseq_args,\n\u001b[1;32m    553\u001b[0m         misprime_lib\u001b[39m=\u001b[39;49mmisprime_lib,\n\u001b[1;32m    554\u001b[0m         mishyb_lib\u001b[39m=\u001b[39;49mmishyb_lib,\n\u001b[1;32m    555\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/envs/ml-s10/lib/python3.11/site-packages/primer3/thermoanalysis.pyx:1289\u001b[0m, in \u001b[0;36mprimer3.thermoanalysis._ThermoAnalysis.run_design\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/miniconda3/envs/ml-s10/lib/python3.11/site-packages/primer3/thermoanalysis.pyx:1592\u001b[0m, in \u001b[0;36mprimer3.thermoanalysis.pdh_design_output_to_dict\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: PRIMER_MAX_END_GC must be between 0 to 5"
     ]
    }
   ],
   "source": [
    "counter = 0\n",
    "for x, row in known_primers.iterrows():\n",
    "    # print(row['Start']-5, row['End'])\n",
    "    sequence = extract_sequence_from_fasta(int(row['Start']-5), int(row['End']), padding = 0, fasta_file= '/mnt/storage10/lwang/Projects/Amplicone_design_tool/MTB-h37rv_asm19595v2-eg18.fa', sequence_id='Chromosome')\n",
    "\n",
    "    middle = extract_sequence_from_fasta(30000, 30500, padding = 0, fasta_file= '/mnt/storage10/lwang/Projects/Amplicone_design_tool/GCF_000002765.5_GCA_000002765_genomic.fna', sequence_id='NC_004325.2')\n",
    "    upper_range = int(row['Amplicon size'])\n",
    "    lower_range = int(row['Amplicon size'] - 10)\n",
    "    # sequence[:250]+'N'*100+sequence[-250:]\n",
    "    # print(sequence)\n",
    "    results = bindings.design_primers(\n",
    "                seq_args={\n",
    "                    'SEQUENCE_ID': 'primer',\n",
    "                    'SEQUENCE_TEMPLATE': sequence,\n",
    "                    'SEQUENCE_INCLUDED_REGION': [0,len(sequence)],\n",
    "                    # 'SEQUENCE_INCLUDED_REGION': [(250,len(sequence)-250)],\n",
    "                    # 'SEQUENCE_EXCLUDED_REGION': [(200,300)],\n",
    "                    \n",
    "                    # 'SEQUENCE_TEMPLATE': sequence[:250]+middle+sequence[-250:],\n",
    "                    # 'SEQUENCE_INCLUDED_REGION': [250,750],\n",
    "                    # 'SEQUENCE_EXCLUDED_REGION': [250,750],\n",
    "\n",
    "                    # 'SEQUENCE_INCLUDED_REGION': sequence[:251],\n",
    "                \n",
    "                },\n",
    "                global_args={\n",
    "                    'PRIMER_MAX_END_GC': 10,\n",
    "                    'PRIMER_PICK_ANYWAY':1,\n",
    "                    'PRIMER_NUM_RETURN': 600,\n",
    "                    # 'PRIMER_OPT_SIZE': 20,\n",
    "                    'PRIMER_PICK_INTERNAL_OLIGO': 1,\n",
    "                    'PRIMER_INTERNAL_MAX_SELF_END': 8,\n",
    "                    # 'PRIMER_MIN_THREE_PRIME_DISTANCE':10,\n",
    "                    # 'PRIMER_MIN_FIVE_PRIME_DISTANCE':10,\n",
    "                    'PRIMER_MIN_SIZE': 16,\n",
    "                    'PRIMER_MAX_SIZE': 25,\n",
    "                    'PRIMER_OPT_TM': 50.0,\n",
    "                    'PRIMER_MIN_TM': 45,\n",
    "                    'PRIMER_MAX_TM': 71,#59.0,\n",
    "                    'PRIMER_MIN_GC': 45,#45.0,\n",
    "                    'PRIMER_MAX_GC': 68,#79.0,\n",
    "                    'PRIMER_MAX_POLY_X': 5,\n",
    "                    'PRIMER_INTERNAL_MAX_POLY_X': 5,\n",
    "                    'PRIMER_SALT_MONOVALENT': 50.0,\n",
    "                    'PRIMER_DNA_CONC': 50.0,\n",
    "                    'PRIMER_MAX_NS_ACCEPTED': 0,\n",
    "                    'PRIMER_MAX_SELF_ANY': 12,\n",
    "                    'PRIMER_MAX_SELF_END': 8,\n",
    "                    'PRIMER_PAIR_MAX_COMPL_ANY': 12,\n",
    "                    'PRIMER_PAIR_MAX_COMPL_END': 8,\n",
    "                    'PRIMER_PRODUCT_SIZE_RANGE': f'{lower_range}-{upper_range}',\n",
    "                    # 'PRIMER_PRODUCT_SIZE_RANGE': '550-650',\n",
    "                    # 'PRIMER_PRODUCT_SIZE_RANGE': [\n",
    "                    #     # [950,1050]\n",
    "                    #     [len(sequence)-350,len(sequence)-250]\n",
    "                    # ],\n",
    "                })\n",
    "    \n",
    "# PRIMER_MIN_TM 45.0\n",
    "# PRIMER_MAX_TM 71\n",
    "# PRIMER_MIN_GC 47\n",
    "# PRIMER_MAX_GC 66\n",
    "    # print(results['PRIMER_PAIR_EXPLAIN'])\n",
    "\n",
    "    # checking primer similarity\n",
    "    query_seq = row['forward']\n",
    "    threshold = 0.99\n",
    "    primer_loc = find_sequence_location(query_seq)\n",
    "    for x in results['PRIMER_LEFT']:\n",
    "        # print(x['SEQUENCE'], x['TM'], x['GC_PERCENT'])\n",
    "        seq_loc = find_sequence_location(x['SEQUENCE'])\n",
    "        # if check_similarity(x['SEQUENCE'], query_seq, threshold=threshold) & seq_loc[0]>primer_loc[0]-10 & seq_loc[1]<primer_loc[1]+10:\n",
    "        if check_similarity(x['SEQUENCE'], query_seq, threshold=threshold):\n",
    "            # checking to see if the primer is within 10 bp of the known primer and have a similarity of that passses the threshold\n",
    "            # print(check_similarity(x['SEQUENCE'], query_seq, threshold=threshold))\n",
    "            \n",
    "            # print(x['SEQUENCE'], x['TM'], x['GC_PERCENT'], x['PENALTY'])\n",
    "            # print(seq_loc)\n",
    "            counter += 1\n",
    "            break\n",
    "print('====================')\n",
    "print(f'{counter}/{known_primers.shape[0]}')\n",
    "print('====================')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "n = 0\n",
    "counter = 0\n",
    "while counter != known_primers.shape[0]:\n",
    "    n += 1\n",
    "    print('>',n,'<')\n",
    "    counter = 0\n",
    "    for x, row in known_primers.iterrows():\n",
    "        # print(row['Start']-5, row['End'])\n",
    "        sequence = extract_sequence_from_fasta(int(row['Start']-5), int(row['End']), padding = 0, fasta_file= '/mnt/storage10/lwang/Projects/Amplicone_design_tool/MTB-h37rv_asm19595v2-eg18.fa', sequence_id='Chromosome')\n",
    "\n",
    "        middle = extract_sequence_from_fasta(30000, 30500, padding = 0, fasta_file= '/mnt/storage10/lwang/Projects/Amplicone_design_tool/GCF_000002765.5_GCA_000002765_genomic.fna', sequence_id='NC_004325.2')\n",
    "        upper_range = int(row['Amplicon size'] + 10)\n",
    "        lower_range = int(row['Amplicon size'] - 10)\n",
    "        # sequence[:250]+'N'*100+sequence[-250:]\n",
    "        # print(sequence)\n",
    "        results = bindings.design_primers(\n",
    "                    seq_args={\n",
    "                        'SEQUENCE_ID': 'primer',\n",
    "\n",
    "                        'SEQUENCE_TEMPLATE': sequence,\n",
    "                        'SEQUENCE_INCLUDED_REGION': [0,len(sequence)],\n",
    "                        # 'SEQUENCE_INCLUDED_REGION': [(250,len(sequence)-250)],\n",
    "                        # 'SEQUENCE_EXCLUDED_REGION': [(200,300)],\n",
    "                        \n",
    "                        # 'SEQUENCE_TEMPLATE': sequence[:250]+middle+sequence[-250:],\n",
    "                        # 'SEQUENCE_INCLUDED_REGION': [250,750],\n",
    "                        # 'SEQUENCE_EXCLUDED_REGION': [250,750],\n",
    "\n",
    "                        # 'SEQUENCE_INCLUDED_REGION': sequence[:251],\n",
    "                    \n",
    "                    },\n",
    "                    global_args={\n",
    "                        'PRIMER_NUM_RETURN': 600,\n",
    "                        # 'PRIMER_OPT_SIZE': 20,\n",
    "                        'PRIMER_PICK_INTERNAL_OLIGO': 0,\n",
    "                        'PRIMER_INTERNAL_MAX_SELF_END': 8,\n",
    "                        # 'PRIMER_MIN_THREE_PRIME_DISTANCE':10,\n",
    "                        # 'PRIMER_MIN_FIVE_PRIME_DISTANCE':10,\n",
    "                        'PRIMER_MIN_SIZE': 17,\n",
    "                        'PRIMER_MAX_SIZE': 24,\n",
    "                        'PRIMER_OPT_TM': 50.0,\n",
    "                        'PRIMER_MIN_TM': 47.0-n,\n",
    "                        'PRIMER_MAX_TM': 69+n,#59.0,\n",
    "                        'PRIMER_MIN_GC': 45-n,#45.0,\n",
    "                        'PRIMER_MAX_GC': 68+n,#79.0,\n",
    "                        'PRIMER_MAX_POLY_X': 5,\n",
    "                        'PRIMER_INTERNAL_MAX_POLY_X': 5,\n",
    "                        'PRIMER_SALT_MONOVALENT': 50.0,\n",
    "                        'PRIMER_DNA_CONC': 500.0,\n",
    "                        'PRIMER_MAX_NS_ACCEPTED': 2,\n",
    "                        'PRIMER_MAX_SELF_ANY': 5,\n",
    "                        'PRIMER_MAX_SELF_END': 2,\n",
    "                        'PRIMER_PAIR_MAX_COMPL_ANY': 5,\n",
    "                        'PRIMER_PAIR_MAX_COMPL_END': 2,\n",
    "                        'PRIMER_PRODUCT_SIZE_RANGE': f'{lower_range}-{upper_range}',\n",
    "                        # 'PRIMER_PRODUCT_SIZE_RANGE': '550-650',\n",
    "                        # 'PRIMER_PRODUCT_SIZE_RANGE': [\n",
    "                        #     # [950,1050]\n",
    "                        #     [len(sequence)-350,len(sequence)-250]\n",
    "                        # ],\n",
    "                    })\n",
    "\n",
    "        # print(results['PRIMER_PAIR_EXPLAIN'])\n",
    "\n",
    "        # checking primer similarity\n",
    "        query_seq = row['forward']\n",
    "        threshold = 0.99\n",
    "        primer_loc = find_sequence_location(query_seq)\n",
    "        for x in results['PRIMER_LEFT']:\n",
    "            # print(x['SEQUENCE'], x['TM'], x['GC_PERCENT'])\n",
    "            seq_loc = find_sequence_location(x['SEQUENCE'])\n",
    "            # if check_similarity(x['SEQUENCE'], query_seq, threshold=threshold) & seq_loc[0]>primer_loc[0]-10 & seq_loc[1]<primer_loc[1]+10:\n",
    "            if check_similarity(x['SEQUENCE'], query_seq, threshold=threshold):\n",
    "                # checking to see if the primer is within 10 bp of the known primer and have a similarity of that passses the threshold\n",
    "                # print(check_similarity(x['SEQUENCE'], query_seq, threshold=threshold))\n",
    "                \n",
    "                # print(x['SEQUENCE'], x['TM'], x['GC_PERCENT'], x['PENALTY'])\n",
    "                # print(seq_loc)\n",
    "                counter += 1\n",
    "                break\n",
    "    print('====================')\n",
    "    print(f'{counter}/{known_primers.shape[0]}')\n",
    "    print('PRIMER_MIN_TM', 47.0-n)\n",
    "    print('PRIMER_MAX_TM', 69+n)#59.0,\n",
    "    print('PRIMER_MIN_GC', 45-n)#45.0,\n",
    "    print('PRIMER_MAX_GC', 68+n)#79.0,\n",
    "\n",
    "    print('====================')\n",
    "\n",
    "print(n)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# checking primer similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6510, 6528)"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_sequence_location('AGAGTTGGTGCGGCGTAA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n",
      "AGAGTTGGTGCGGCGTAA 62.88089938584807 55.55555555555556\n",
      "(6510, 6528)\n"
     ]
    }
   ],
   "source": [
    "# checking primer similarity\n",
    "query_seq = 'AGAGTTGGTGCGGCGTAA'\n",
    "threshold = 0.99\n",
    "primer_loc = [6510, 6928]\n",
    "for x in results['PRIMER_LEFT']:\n",
    "    # print(x['SEQUENCE'], x['TM'], x['GC_PERCENT'])\n",
    "    seq_loc = find_sequence_location(x['SEQUENCE'])\n",
    "    # if check_similarity(x['SEQUENCE'], query_seq, threshold=threshold) & seq_loc[0]>primer_loc[0]-10 & seq_loc[1]<primer_loc[1]+10:\n",
    "    if check_similarity(x['SEQUENCE'], query_seq, threshold=threshold):\n",
    "        # checking to see if the primer is within 10 bp of the known primer and have a similarity of that passses the threshold\n",
    "        # print(check_similarity(x['SEQUENCE'], query_seq, threshold=threshold))\n",
    "        print(x['SEQUENCE'], x['TM'], x['GC_PERCENT'])\n",
    "        print(seq_loc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_similarity('ATCGATCGATCG', 'ATCGATCGATCG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'internal_oligo_complementarity': ThermoResult(structure_found=True, tm=35.91, dg=-10909.11, dh=-86800.00, ds=-244.69),\n",
       " 'consecutive_identical_bases': ThermoResult(structure_found=True, tm=50.69, dg=-1174.97, dh=-27800.00, ds=-85.85)}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# from Bio.Emboss.primer3 import Primer3\n",
    "def calculate_primer_properties(primer_sequence):\n",
    "    primer_properties = {}\n",
    "    # Calculate 3' self-complementarity of an internal oligo\n",
    "    primer_properties['internal_oligo_complementarity'] = primer3.calcHomodimer(primer_sequence)\n",
    "    # Calculate the size of consecutive identical base repeats in a primer\n",
    "    primer_properties['consecutive_identical_bases'] = primer3.calcHairpin(primer_sequence)\n",
    "    # Calculate the self-complementarity score for any part of a primer\n",
    "    # primer_properties['any_part_complementarity'] = primer3.calcHeterodimer(primer_sequence)\n",
    "    # Calculate the self-complementarity score for the 3' end of a primer\n",
    "    # primer_properties['three_prime_complementarity'] = primer3.calcEndStability(primer_sequence)\n",
    "    return primer_properties\n",
    "\n",
    "calculate_primer_properties('ATCGATCGATCG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml-s10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
